include/gatekeeper_log_ratelimit.h

```diff
@@ -0,0 +1,109 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_LOG_RATELIMIT_H_
+#define _GATEKEEPER_LOG_RATELIMIT_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <rte_memory.h>
+
+/*
+ * The code of this file is mostly a copy of the Linux kernel.
+ */
+
+struct log_ratelimit_state {
+	uint32_t ratelimit_level;
+	uint64_t interval_cycles;
+	uint32_t burst;
+	uint32_t printed;
+	uint64_t begin;
+} __rte_cache_aligned;
+
+ /*
+  * @lcore_id: initialize the log_ratelimit_state data for @lcore_id.
+  *
+  * This will allow to only enforce a rate limit on log entries
+  * with a level higher or equal than @log_ratelimit_state.ratelimit_level:
+  * no more than @log_ratelimit_state.burst callbacks in
+  * every @log_ratelimit_state.interval milliseconds.
+  *
+  * Note that, to avoid performance degradation caused by locks, the
+  * implementation assumes that each lcore will maintain a separate
+  * log_ratelimit_state data to rate limit the log rates.
+  */
+void log_ratelimit_state_init(unsigned lcore_id,
+	uint32_t ratelimit_level, uint32_t interval, uint32_t burst);
+bool log_ratelimit_allow(struct log_ratelimit_state *lrs);
+
+/**
+ * Generates and ratelimits a log message.
+ *
+ * The message will be sent in the stream defined by the previous call
+ * to rte_openlog_stream() if it is not ratelimited.
+ *
+ * The level argument determines if the log should be displayed or
+ * not, depending on the global rte_logs variable.
+ *
+ * The preferred alternative is the RTE_LOG_RATELIMIT() because it adds the
+ * level and type in the logged string.
+ *
+ * @param level
+ *   Log level. A value between RTE_LOG_EMERG (1) and RTE_LOG_DEBUG (8).
+ * @param logtype
+ *   The log type, for example, RTE_LOGTYPE_EAL.
+ * @param format
+ *   The format string, as in printf(3), followed by the variable arguments
+ *   required by the format.
+ * @return
+ *   - 0: Success.
+ *   - Negative on error.
+ */
+int rte_log_ratelimit(uint32_t level, uint32_t logtype, const char *format, ...)
+#ifdef __GNUC__
+#if (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ > 2))
+	__attribute__((cold))
+#endif
+#endif
+	__attribute__((format(printf, 3, 4)));
+
+/**
+ * Generates and ratelimits a log message.
+ *
+ * The RTE_LOG_RATELIMIT() is a helper that prefixes the string with
+ * the log level and type, and call rte_log_ratelimit().
+ *
+ * @param l
+ *   Log level. A value between EMERG (1) and DEBUG (8). The short name is
+ *   expanded by the macro, so it cannot be an integer value.
+ * @param t
+ *   The log type, for example, EAL. The short name is expanded by the
+ *   macro, so it cannot be an integer value.
+ * @param ...
+ *   The fmt string, as in printf(3), followed by the variable arguments
+ *   required by the format.
+ * @return
+ *   - 0: Success.
+ *   - Negative on error.
+ */
+#define RTE_LOG_RATELIMIT(l, t, ...)                     \
+	rte_log_ratelimit(RTE_LOG_ ## l,                 \
```

Comment:
Macro `RTE_LOG_RATELIMIT()` is never used. Moreover, we don't want to use it in Gatekeeper since each block has its own macro. Let's drop it.